### PR TITLE
Integrate ResourceNames in DiscoGapic generated surface

### PR DESCRIPTION
**Pull request by andreamlin**
- Generate flattened methods that use the ResourceName as a parameter, instead of the multiple Strings that make up the URI.
- The name of the ResourceName class is singular, e.g. "AddressName" instead of "AddressesName".

### CODE REVIEW

**Comment by andreamlin** [comment]
PTAL.

File: src/main/java/com/google/api/codegen/DiscoGapicGeneratorApi.java [hunk 1]
```diff
@@ -48,8 +67,10 @@
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import org.apache.commons.lang3.NotImplementedException;
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/DiscoGapicGeneratorApi.java]: I prefer UnsupportedOperationException (since it's part of the JDK)

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline [hunk 2]
```diff
@@ -4517,6 +4657,33 @@ public class AddressAdminClient implements BackgroundResource {
    * }
    * </code></pre>
    *
+   * @param region Name of the region for this request.
+   * @param address A reserved address resource.
+   * @throws com.google.api.gax.rpc.ApiException if the remote call fails
+   */
+  @BetaApi
+  public final Operation addressesInsert(RegionName region, Address address) {
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: Comment for later work: to match our existing API naming patterns, this should be `insertAddresses`

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline [hunk 3]
```diff
@@ -4517,6 +4657,33 @@ public class AddressAdminClient implements BackgroundResource {
...
+
+  // AUTO-GENERATED DOCUMENTATION AND METHOD
+  /**
+   * Creates an address resource in the specified project using the data included in the request.
+   *
+   * Sample code:
+   * <pre><code>
+   * try (AddressAdminClient addressAdminClient = AddressAdminClient.create()) {
+   *
+   *   Operation response = addressAdminClient.addressesInsert();
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: The sample method call here needs to pass a request

File: src/main/java/com/google/api/codegen/DiscoGapicGeneratorApi.java [hunk 4]
```diff
@@ -84,6 +105,18 @@
 
   private final ToolOptions options;
 
+  private static final Map<String, Class<? extends SurfaceNamer>> SURFACE_NAMER_MAP =
```
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/DiscoGapicGeneratorApi.java]: removed.

File: src/main/java/com/google/api/codegen/DiscoGapicGeneratorApi.java (hunk 1 above)
Comment (andreamlin) [inline src/main/java/com/google/api/codegen/DiscoGapicGeneratorApi.java]: Done.

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline (hunk 2 above)
Comment (andreamlin) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: a couple lines did the job

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline (hunk 3 above)
Comment (andreamlin) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: There is a TODO in InitCodeTransformer for filling out the parameters of the sample method call. It's more work than I'd like to get into right now.

**Comment by andreamlin** [comment]
Addressed comments.
PTAL.

File: src/main/resources/com/google/api/codegen/java/message.snip [hunk 5]
```diff
@@ -112,7 +112,7 @@
   }
   private static final {@schema.innerTypeName} DEFAULT_INSTANCE;
   static {
-    DEFAULT_INSTANCE = new {@schema.innerTypeName}();
+      DEFAULT_INSTANCE = new {@schema.innerTypeName}();
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/message.snip]: This should be 2-space indentation instead of 4

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline [hunk 6]
```diff
@@ -4624,16 +4791,45 @@ public class AddressAdminClient implements BackgroundResource {
...
    * }
    * </code></pre>
    *
    * @param request The request object containing all of the parameters for the API call.
    * @throws com.google.api.gax.rpc.ApiException if the remote call fails
    */
   @BetaApi
-  private final Address addressesGet(AddressesGetHttpRequest request) {
-    return addressesGetCallable().call(request);
+  private final Address getAddresses(AddressesGetHttpRequest request) {
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: It looks like there is one more level of name manipulation that needs to be done. The CRUD operations should be in singular (`getAddress`, `insertAddress`, `deleteAddress`) but the list operation should remain in plural (`listAddresses`). We have a method for singularization in toolkit already (but it is kind of incomplete): https://github.com/googleapis/toolkit/blob/master/src/main/java/com/google/api/codegen/Inflector.java

Also, the request type name needs to be reordered too.

File: src/main/resources/com/google/api/codegen/java/message.snip (hunk 5 above)
Comment (andreamlin) [inline src/main/resources/com/google/api/codegen/java/message.snip]: I'm not sure how else to make the baseline indentation look nice. With just one tab, the resulting view is
```
static {
DEFAULT_INSTANCE
}
```

but the two tab indentation of the latest revision results in
```
static {
  DEFAULT_INSTANCE
}
```

My guess is that the curly braces mess with the snippet parsing whitespacing.

Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/message.snip]: Oh, the rest of them look fine; this particular one was already correct, but it got indented too far.

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline (hunk 6 above)
Comment (andreamlin) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: Fixed the ordering of the request type names.

Applied singularization to non-plural operations' names. I added a couple more cases to Inflector; hopefully English's corner cases won't be much of a problem later on.

File: src/main/resources/com/google/api/codegen/java/message.snip (hunk 5 above)
Comment (andreamlin) [inline src/main/resources/com/google/api/codegen/java/message.snip]: So what should be done to fix this? Without the new extra indentation, the resulting view doesn't have an indentation.

**Comment by andreamlin** [comment]
Responded to all comments.
PTAL.

File: src/main/resources/com/google/api/codegen/java/message.snip (hunk 5 above)
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/message.snip]: Wait you said "tab"... it should only be indented with spaces.

File: src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline (hunk 6 above)
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/discogapic/java/java_simplecompute.v1.json.baseline]: There are quite a lot of corner cases in English... I did a check of several singularization algorithms on the web, and all of them failed on something. A fully correct singularizer needs to have a bunch of cases for specific words (compare: drives -> drive, lives -> life).

File: src/main/resources/com/google/api/codegen/java/message.snip (hunk 5 above)
Comment (andreamlin) [inline src/main/resources/com/google/api/codegen/java/message.snip]: whoops, by "tab" I meant to say "two spaces"
